Lease Checkpoints fix

serathius opened the pull request:
This is continuation of work on https://github.com/etcd-io/etcd/pull/13491
Main difference is that this proposes to persist remainingTTL on checkpoint making checkpoints not depend on snapshot frequency.

I also check that renew logic should not be impacted by this change. Renew already includes logic that will reset remainingTTL if needed.

serathius commented:
cc @ptabor

serathius commented:
cc @jpbetz

ptabor commented:
Can this cause inconsistency of checksums of the DBs computed from diffrent instances, 
i.e. checkpointing is enabled with 3.6 (or 3.5.3 backport) so version 3.5.2 would had different post-raft state in comparison to version that didn't got upgraded to 3.5.2... 
We should have tests / know what are the implications of such 'mismatch',

serathius commented:
I have added a proper feature guarding to ensure that we avoid checksum inconsistency.
First we require cluster version to be at least v3.6, this way will handle upgrades as cluster version is bumped only after all members of cluster are upgraded. Second I added a flag to provide compatibility for v3.5 -> v3.6 upgrades. As we are planning to backport this fix to v3.5 this flag will be used in v3.5 to enable checkpoint persist. However for v3.6 this flag will be used to ensure consistency during upgrade as it will inform v3.6 members that cluster already has this feature enabled.

ptabor commented on server/etcdmain/config.go:
```diff
@@ -282,7 +282,8 @@ func newConfig() *config {
 
-	fs.BoolVar(&cfg.ec.ExperimentalEnableLeaseCheckpoint, "experimental-enable-lease-checkpoint", false, "Enable to persist lease remaining TTL to prevent indefinite auto-renewal of long lived leases.")
+	fs.BoolVar(&cfg.ec.ExperimentalEnableLeaseCheckpoint, "experimental-enable-lease-checkpoint", false, "Enable leader to send regular checkpoints to other members to prevent reset of remaining TTL on leader change.")
+	fs.BoolVar(&cfg.ec.ExperimentalLeaseCheckpointPersist, "experimental-enable-lease-checkpoint-persist", false, "Enable persisting remainingTTL to prevent indefinite auto-renewal of long lived leases. Always enabled in v3.6. Should be used to ensure smooth upgrade from v3.5 clusters with this feature enabled.")
```
maybe: ExperimentalEnableLeaseCheckpointPersist  to be consistent with the flag name. 

// TODO to delete it in 3.7.

serathius replied:
Done.

ptabor commented on server/lease/lessor.go:
```diff
@@ -351,6 +366,10 @@ func (le *lessor) Checkpoint(id LeaseID, remainingTTL int64) error {
 		// when checkpointing, we only update the remainingTTL, Promote is responsible for applying this to lease expiry
 		l.remainingTTL = remainingTTL
+		cv := le.cluster.Version()
+		if le.checkpointPersist || (cv != nil && greaterOrEqual(*cv, v3_6)) {
```
nit: shell we have this embedded in a getter on `le`. 

le.isCheckpointPersist() that hides the v3.6 resolution ?

serathius replied:
SG

ptabor commented on server/lease/lessor.go:
```diff
@@ -359,6 +378,10 @@ func (le *lessor) Checkpoint(id LeaseID, remainingTTL int64) error {
 }
 
+func greaterOrEqual(first, second semver.Version) bool {
+	return first.Equal(second) || second.LessThan(first)
```
Is it equivalent of: `!(second.LessThen(first))` or there is something subtle ?

serathius replied:
Nothing subtle, fixed.

ptabor commented on server/lease/lessor_test.go:
```diff
@@ -580,6 +581,75 @@ func TestLessorCheckpointsRestoredOnPromote(t *testing.T) {
+				t.Errorf("remainingTTL() = %d, expected: 10", l.RemainingTTL())
+			}
+			le.Checkpoint(2, checkpointTTL)
+			if l.RemainingTTL() != 5 {
```
I think we should use constant's here as well... otherwise it's not clear why we do expect '5'

serathius replied:
Done

ptabor commented on server/lease/lessor_test.go:
```diff
@@ -580,6 +581,75 @@ func TestLessorCheckpointsRestoredOnPromote(t *testing.T) {
+				t.Fatal(err)
+			}
+			if l.RemainingTTL() != ttl {
+				t.Errorf("remainingTTL() = %d, expected: 10", l.RemainingTTL())
```
we should print value of `ttl` instead of hardcoded 10.

ptabor reviewed (approved):
LGTM. Thank you.

jpbetz reviewed (commented):
The use of persistTo and scheduling checkpoints on promotion look right to me. Thanks for fixing. One comment about flag docs.

serathius commented on server/lease/lessor_test.go:
```diff
@@ -580,6 +581,75 @@ func TestLessorCheckpointsRestoredOnPromote(t *testing.T) {
 }
 
+func TestLessorCheckpointPersistenceAfterRestart(t *testing.T) {
+	var ttl int64 = 10
```
Done

ptabor commented on server/embed/config.go:
```diff
@@ -315,9 +315,13 @@ type Config struct {
+	// ExperimentalEnableLeaseCheckpoint enables leader to send regular checkpoints to other members to prevent reset of remaining TTL on leader change.
 	ExperimentalEnableLeaseCheckpoint bool `json:"experimental-enable-lease-checkpoint"`
-	ExperimentalCompactionBatchLimit  int  `json:"experimental-compaction-batch-limit"`
+	// ExperimentalLeaseCheckpointPersist enables persisting remainingTTL to prevent indefinite auto-renewal of long lived leases. Always enabled in v3.6. Should be used to ensure smooth upgrade from v3.5 clusters with this feature enabled.
```
+1. Maybe we should add on startup:
  a) log-warning: if LeaseCheckpoint is enabled, LeaseCheckpointPersistence should be enabled as well. 
  b) fail: if LeaseCheckpoint is disabled & LeaseCheckpointPersistence is set.

serathius replied:
Done

serathius commented:
Addressed all the comments, @ptabor please take another look.

ahrtr commented:
@serathius  Do you have bandwidth to backport this PR to 3.4? :)

serathius commented:
Done
